presto-main/src/main/java/com/facebook/presto/metadata/FunctionUtils.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.metadata;
+
+import com.facebook.presto.spi.function.OperatorType;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.spi.type.VarcharType;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.primitives.Primitives;
+import io.airlift.slice.Slice;
+
+import java.util.Set;
+
+import static com.facebook.presto.metadata.FunctionKind.SCALAR;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+import static com.facebook.presto.spi.type.VarbinaryType.VARBINARY;
+import static com.google.common.base.Preconditions.checkArgument;
+
+public final class FunctionUtils
+{
+    // hack: java classes for types that can be used with magic literals
+    private static final Set<Class<?>> SUPPORTED_LITERAL_TYPES = ImmutableSet.of(long.class, double.class, Slice.class, boolean.class);
+    static final String MAGIC_LITERAL_FUNCTION_PREFIX = "$literal$";
+    private static final String OPERATOR_PREFIX = "$operator$";
+
+    private FunctionUtils() {}
+
+    public static Signature getMagicLiteralFunctionSignature(Type type)
```

Comment:
Ok did a bunch of digging...  I suggest we move the *literal* related functions to `LiteralEncoder`.  I also suggest we hide the "magic" literal stuff a bit better, by introducing this method:
```
    private static Optional<Signature> resolveLiteralFunction(QualifiedName name, List<TypeSignatureProvider> parameterTypes, TypeManager typeManager)
    {
        if (!name.getSuffix().startsWith(MAGIC_LITERAL_FUNCTION_PREFIX)) {
            return Optional.empty();
        }

        // extract type from function name
        String typeName = name.getSuffix().substring(MAGIC_LITERAL_FUNCTION_PREFIX.length());

        // lookup the type
        Type type = typeManager.getType(parseTypeSignature(typeName));

        // verify we have one parameter of the proper type
        checkArgument(parameterTypes.size() == 1, "Expected one argument to literal function, but got %s", parameterTypes);

        return Optional.of(getMagicLiteralFunctionSignature(type));
    }
```

Then in `FunctionRegistry.resolveFunction`, we do something like:
```
        return resolveLiteralFunction(name, parameterTypes, typeManager)
                .orElseThrow(() -> new PrestoException(FUNCTION_NOT_FOUND, message));
```

Reply:
If this is just a private function in `FunctionRegistry` not sure how much "hiding" it adds. A practical problem with this is that there's an error about `message` being not `final`. Do you think this is really necessary?